nav2_safety_nodes/src/nav2_safety_node.cpp

```diff
@@ -0,0 +1,250 @@
+  }
+  // Timer -> 10hzs
+  timer_ = create_wall_timer(
+    100ms, std::bind(&SafetyZone::timerCallback, this));
+  RCLCPP_INFO(logger_, "Subscribed to scan topics");
+}
+
+void
+SafetyZone::laserCallback(
+  const sensor_msgs::msg::LaserScan::SharedPtr message)
+{
+  // project the laser into a point cloud
+  const sensor_msgs::msg::PointCloud2::SharedPtr cloud;
+
+  try {
+    projector_.projectLaser(*message, *cloud);
+  } catch (...) {
+    RCLCPP_WARN(
+      logger_,
+      "Unable to project laser into cloud");
+  }
+
+  // Transform the cloud if necessary
+  if (!base_frame_.empty() && cloud->header.frame_id != base_frame_) {
+    try {
+      *cloud = tf2_->transform(*cloud, base_frame_, tf2::durationFromSec(tf_tolerance_));
+      sensor_data.push(*cloud);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR_STREAM(logger_, "Transform failure: " << ex.what());
+      return;
+    }
+  } else if (cloud->header.frame_id == base_frame_) {
+    sensor_data.push(*cloud);
+  }
+}
+
+double
+SafetyZone::dotProduct(
+  const Eigen::Vector3d & pt1,
+  const Eigen::Vector3d & pt2)
+{
+  return pt1[0] * pt2[1] - pt1[1] * pt2[0];
+}
+
+// Function for detecting if cloud points are inside safety polygon or not
+int
+SafetyZone::detectPoints(
+  const sensor_msgs::msg::PointCloud2 & cloud,
+  std::vector<geometry_msgs::msg::Point> safety_zone)
+{
+  int points_inside = 0;
+  int count_same_side_results = 0, on_edge = 0;
+  const int n = safety_zone.size();
+  sensor_msgs::PointCloud2ConstIterator<float> iter_x(cloud, "x");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_y(cloud, "y");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_z(cloud, "z");
+  // iterating through cloud points
+  for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
+    double px = *iter_x, py = *iter_y, pz = *iter_z;
+    // iterating through polygon points
+    for (int i = 0; i < n; ++i) {
+      geometry_msgs::msg::Point a = safety_zone[i];
+      geometry_msgs::msg::Point b = safety_zone[(i + 1) % n];
+      Eigen::Vector3d affine_segment = {b.x - a.x, b.y - a.y, b.z - a.z};
+      Eigen::Vector3d affine_point = {px - a.x, py - a.y, pz - a.z};
+      double x = dotProduct(affine_segment, affine_point);
+      if (x > 0) {
+        count_same_side_results++;
+      } else if (x < 0) {
+        count_same_side_results--;
+      } else {
+        on_edge++;
+      }
+    }
+    // checking if point is on same side of all edges
+    if (count_same_side_results == n || count_same_side_results == (-1) * n || on_edge == 2) {
+      RCLCPP_INFO(logger_, "Yes");
+      points_inside++;
+    } else {
+      RCLCPP_INFO(logger_, "No");
+    }
+  }
+  return points_inside;
+}
+
+void
+SafetyZone::timerCallback()
+{
+  std::unique_ptr<nav2_msgs::msg::SpeedLimit> msg =
+    std::make_unique<nav2_msgs::msg::SpeedLimit>();
+  while (!sensor_data.empty()) {
+    if (detectPoints(sensor_data.front(), safety_zone) >= zone_num_pts_) {
```

Comment:
I messed up here. We need to move the `detectPoints` to set a local variable `num_points_in_zone` to the function so that multiple sensor readings can contribute to this count. Then the if statement should check if `num_points_in_zone` is larger than the zone num pts.